File: src/go/scanner/scanner.go
package scanner

import (
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"virtual-tv/src/go/config"
)

func GetDuration(path string) (float64, error) {
	cmd := exec.Command("ffprobe", "-v", "error", "-show_entries", "format=duration", "-of", "default=noprint_wrappers=1:nokey=1", path)
	out, err := cmd.Output()
	if err != nil {
		return 0, err
	}
	durationStr := strings.TrimSpace(string(out))
	return strconv.ParseFloat(durationStr, 64)
}

func FullScan(shows []string) {
	var episodes []config.Episode

	for _, showPath := range shows {
		showName := filepath.Base(showPath)
		filepath.Walk(showPath, func(path string, info os.FileInfo, err error) error {
			if err != nil {
				return nil
			}
			if !info.IsDir() {
				ext := strings.ToLower(filepath.Ext(path))
				if ext == ".mp4" || ext == ".mkv" || ext == ".avi" || ext == ".mov" || ext == ".webm" {
					duration, err := GetDuration(path)
					if err == nil {
						episodes = append(episodes, config.Episode{
							Path:     path,
							Name:     strings.TrimSuffix(info.Name(), filepath.Ext(info.Name())),
							Show:     showName,
							Duration: duration,
						})
					}
				}
			}
			return nil
		})
	}

	config.WriteLibrary(config.Library{Episodes: episodes})
}

File: src/backend/scanner.js
const fs = require('fs')
const path = require('path')
const { execSync } = require('child_process')
const { readLibrary, writeLibrary } = require('./config')

const VIDEO_EXTENSIONS = ['.mp4', '.mkv', '.avi', '.mov', '.webm']

function getVideoDuration (filePath) {
  try {
    const cmd = `ffprobe -v error -show_entries format=duration -of default=noprint_wrappers=1:nokey=1 "${filePath}"`
    const output = execSync(cmd).toString().trim()
    return isNaN(parseFloat(output)) ? null : parseFloat(output)
  } catch (e) {
    return null
  }
}

function scanFolder (dir, library) {
  const files = fs.readdirSync(dir)
  const results = []

  for (const file of files) {
    const fullPath = path.join(dir, file)
    const stat = fs.statSync(fullPath)

    if (stat.isDirectory()) {
      results.push(...scanFolder(fullPath, library))
    } else {
      const ext = path.extname(file).toLowerCase()
      if (VIDEO_EXTENSIONS.includes(ext)) {
        // Check if we already have it in library and have duration
        let duration = null
        // Flattened lookup in library for performance? 
        // For simplicity, we'll just search.
        for (const showPath in library) {
          const found = library[showPath].find(ep => ep.path === fullPath)
          if (found && found.duration) {
            duration = found.duration
            break
          }
        }

        if (!duration) {
          console.log(`Probing duration for ${fullPath}...`)
          duration = getVideoDuration(fullPath)
        }

        results.push({
          name: file,
          path: fullPath,
          duration: duration || 22 * 60 // Fallback to 22 mins
        })
      }
    }
  }
  return results
}

function fullScan (shows) {
  const library = readLibrary()
  const newLibrary = {}

  for (const showPath of shows) {
    if (fs.existsSync(showPath)) {
      console.log(`Scanning ${showPath}...`)
      newLibrary[showPath] = scanFolder(showPath, library)
    }
  }

  writeLibrary(newLibrary)
  return newLibrary
}

module.exports = {
  fullScan
}
